File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        Arrays.fill(answer, -1);
        Deque<Integer> stack = new LinkedList<>();
        
        for (int i = 0; i < numbers.length; i++) {
            while(!stack.isEmpty() && numbers[stack.peekLast()] < numbers[i]) {
                answer[stack.pollLast()] = numbers[i];
            }
            stack.add(i);
            
        }
        
        
        
        
        return answer;
    }
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
from collections import deque

def solution(numbers):
    answer = [-1] * len(numbers)
    q = deque()
    for i in range(len(numbers)):    
        while q and q[-1][0] < numbers[i]:
            value, index = q.pop()
            answer[index] = numbers[i]
        
        q.append((numbers[i], i))
    
    return answer